File: AiSD1/fifo.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>

typedef struct Element Element;
typedef struct Queue Queue;

struct Element {
    int value;
    Element* prev;
};

struct Queue {
    Element* head;
    Element* tail;
    int size;
};

Queue* createQueue () {
    Queue* queue = malloc (sizeof (Queue));
    queue -> head = NULL;
    queue -> tail = NULL;
    queue -> size = 0;
    return queue;
}

bool isEmpty (Queue* queue) {
    if (queue -> size == 0) {
        return true;
    } else {
        return false;
    }
}

void initQueue (Queue* queue, int value) {
    Element* element = malloc (sizeof (Element));
    element -> value = value;
    element -> prev = NULL;
    queue -> head = element;
    queue -> tail = element;
    queue -> size = 1;
}

void addToQueue (Queue* queue, int value) {

    if (isEmpty(queue)) {
        initQueue (queue, value);
        return; 
    }

    Element* element = malloc (sizeof (Element));
    element -> value = value;
    element -> prev = NULL;

    queue -> tail -> prev = element;
    queue -> tail = element;
    queue -> size++;
}

bool takeFromQueue (Queue* queue) {
    if (isEmpty(queue)) {
        return false;
    }

    Element* tmp = queue -> head -> prev;
    free (queue -> head);
    queue -> head = tmp;
    queue -> size--;
    return true;
}

void showQueue (Queue* queue) {
    if (isEmpty(queue)) {
        printf("EMPTY!\n");
        return;
    }
    Element* tmp;
    tmp = queue -> head;
    while (tmp != NULL) {
        printf ("%d <- ", tmp -> value);
        tmp = tmp -> prev;
    }
    printf ("NULL\n");
}

int main () {
    Queue* queue = createQueue();
    addToQueue (queue, 16);
    addToQueue (queue, 54);
    addToQueue (queue, 98);
    addToQueue (queue, 17);
    showQueue(queue);
    takeFromQueue (queue);
    showQueue(queue);
    takeFromQueue (queue);
    showQueue(queue);
    takeFromQueue (queue);
    showQueue(queue);
    takeFromQueue (queue);
    showQueue(queue);

    int value;
    Queue* test = createQueue();
    srand(time(NULL));
    for (int i = 0; i < 100; i++) {
        value = rand() % 100;
        addToQueue (test, value);
    }
    showQueue(test);
    for (int i = 0; i < 10; i++) {
        takeFromQueue(test);
        printf("\n");
        showQueue(test);
        
    }

    return 0;
}

File: AiSD1/L1/go/fifo.go
package main

import "fmt"

type Element struct {
	value int
	prev  *Element
}

type Queue struct {
	head *Element
	tail *Element
	size int
}

func createQueue() *Queue {
	queue := &Queue{head: nil, tail: nil, size: 0}
	return queue
}

func isEmpty(q *Queue) bool {
	return q.size == 0
}

func initQueue(q *Queue, val int) {
	element := &Element{value: val, prev: nil}
	q.head = element
	q.tail = element
	q.size = 1
}

func insert(q *Queue, val int) {
	if isEmpty(q) {
		initQueue(q, val)
	}

	e := &Element{prev: nil, value: val}
	q.tail.prev = e
	q.tail = e
	q.size++
}

func delete(q *Queue) bool {
	if isEmpty(q) {
		return false
	}

	tmp := q.head.prev
	q.head = tmp
	q.size--
	return true
}

func print(q *Queue) {
	if isEmpty(q) {
		fmt.Println("Queue is empty")
		return
	}


	tmp := q.head
	for tmp != nil {
		fmt.Print(tmp.value, "<- ")
		tmp = tmp.prev
	}

	fmt.Println("NIL")
}



func main() {
	queue := createQueue()
	insert(queue, 16)
	insert(queue, 54)
    insert(queue, 98)
    insert(queue, 17)
	print(queue)
	delete(queue)
	print(queue)
	delete(queue)
	print(queue)
	delete(queue)
	print(queue)
	delete(queue)
}
